File: docs/examples/python-lsp/actions.py
"""Code-actions playground (Alt-Q A).

Python's code actions are the most server-dependent of the wired languages:

* pyright offers "Organize Imports" and quick-fixes attached to diagnostics
  (e.g. add a missing import).
* pylsp offers refactors -- "Extract variable/method", "Inline" -- only when the
  optional rope plugin is installed (python3-rope), and quick-fixes from
  pyflakes/pycodestyle when those are installed.

Put the cursor on a marked spot, run Alt-Q A, pick an entry; the buffer is
rewritten in place (F2 saves, Ctrl-U undoes, Ctrl-R redoes).
"""

import os       # deliberately unused -> Alt-Q A on the import line: Organize Imports
import sys       # (unused on purpose)


# Alt-Q A on the `"hi " + who` expression (pylsp + rope) -> "Extract variable".
def greet(who):
    return "hi " + who          # Alt-Q A here -> extract variable (rope)


# Alt-Q A on the body (pylsp + rope) -> "Extract method".
def area_of(width, height):
    return width * height        # Alt-Q A here -> extract method (rope)

File: docs/examples/rust-lsp/src/actions.rs
//! Code-actions playground (Alt-Q A).  rust-analyzer ships a rich set of
//! "assists" (refactors) plus quick-fixes attached to compiler diagnostics.
//! Put the cursor on a marked spot, run Alt-Q A, pick an entry; the buffer is
//! rewritten in place (F2 saves, Ctrl-U undoes, Ctrl-R redoes).

#![allow(dead_code)]

// Alt-Q A on the `"hi ".to_string() + who` expression -> "Replace with format!".
fn greet(who: &str) -> String {
    "hi ".to_string() + who // Alt-Q A here -> convert to format!()
}

// Alt-Q A on the `if` -> assists like "Convert to match" / "Invert if".
fn classify(n: i32) -> i32 {
    if n > 0 {
        1
    } else {
        helper(n)
    } // Alt-Q A on `if` -> convert to match
}

// `unused` is never read -> rust-analyzer flags it (Alt-Q E); Alt-Q A offers a fix.
fn helper(n: i32) -> i32 {
    let unused = 42; // Alt-Q A here -> remove this / prefix with underscore
    n * 2
}
